orderer/consensus/smartbft/chain.go

```diff
@@ -0,0 +1,604 @@
+	if len(blockClone.Metadata.Metadata) > int(cb.BlockMetadataIndex_SIGNATURES) {
+		signatureMetadata := &cb.Metadata{}
+		// Nil out signatures because we carry them around separately in the library format.
+		if err := proto.Unmarshal(blockClone.Metadata.Metadata[cb.BlockMetadataIndex_SIGNATURES], signatureMetadata); err != nil {
+			// nothing to do
+			c.Logger.Errorf("Error unmarshalling signature metadata from block: %s", err)
+		}
+		signatureMetadata.Signatures = nil
+		blockClone.Metadata.Metadata[cb.BlockMetadataIndex_SIGNATURES] = protoutil.MarshalOrPanic(signatureMetadata)
+	}
+	prop := types.Proposal{
+		Header: protoutil.BlockHeaderBytes(blockClone.Header),
+		Payload: (&ByteBufferTuple{
+			A: protoutil.MarshalOrPanic(blockClone.Data),
+			B: protoutil.MarshalOrPanic(blockClone.Metadata),
+		}).ToBytes(),
+		VerificationSequence: int64(c.verifier.VerificationSequence()),
+	}
+
+	if protoutil.IsConfigBlock(block) {
+		prop.VerificationSequence--
+	}
+
+	return prop
+}
+
+func (c *BFTChain) blockToDecision(block *cb.Block) *types.Decision {
+	proposal := c.blockToProposalWithoutSignaturesInMetadata(block)
+	if block.Header.Number == 0 {
+		return &types.Decision{
+			Proposal: proposal,
+		}
+	}
+
+	signatureMetadata := &cb.Metadata{}
+	if err := proto.Unmarshal(block.Metadata.Metadata[cb.BlockMetadataIndex_SIGNATURES], signatureMetadata); err != nil {
+		c.Logger.Panicf("Failed unmarshaling signatures from block metadata: %v", err)
+	}
+
+	ordererMDFromBlock := &cb.OrdererBlockMetadata{}
+	if err := proto.Unmarshal(signatureMetadata.Value, ordererMDFromBlock); err != nil {
+		c.Logger.Panicf("Failed unmarshaling OrdererBlockMetadata from block signature metadata: %v", err)
+	}
+
+	proposal.Metadata = ordererMDFromBlock.ConsenterMetadata
+
+	var signatures []types.Signature
+	for _, sigMD := range signatureMetadata.Signatures {
+		idHdr := &cb.IdentifierHeader{}
+		if err := proto.Unmarshal(sigMD.IdentifierHeader, idHdr); err != nil {
+			c.Logger.Panicf("Failed unmarshaling identifier header for %s", base64.StdEncoding.EncodeToString(sigMD.IdentifierHeader))
+		}
+		sig := &Signature{
+			IdentifierHeader:     sigMD.IdentifierHeader,
+			BlockHeader:          protoutil.BlockHeaderBytes(block.Header),
+			OrdererBlockMetadata: signatureMetadata.Value,
+		}
+		signatures = append(signatures, types.Signature{
+			Msg:   sig.Marshal(),
+			Value: sigMD.Signature,
+			ID:    uint64(idHdr.Identifier),
+		})
+	}
+
+	return &types.Decision{
+		Signatures: signatures,
+		Proposal:   proposal,
+	}
+}
+
+// HandleMessage handles the message from the sender
+func (c *BFTChain) HandleMessage(sender uint64, m *smartbftprotos.Message) {
+	c.Logger.Debugf("Message from %d", sender)
+	c.consensus.HandleMessage(sender, m)
+}
+
+// HandleRequest handles the request from the sender
+func (c *BFTChain) HandleRequest(sender uint64, req []byte) {
+	c.Logger.Debugf("HandleRequest from %d", sender)
+	c.consensus.SubmitRequest(req)
```

Comment:
we need to verify the request. Otherwise a byzantine node may send us a request and we will enqueue it in the block and the followers will detect it, and will start a view change.